File: src/cache.ts
import { createHash } from 'node:crypto'
import { promises as fs } from 'node:fs'
import path from 'node:path'
import { spawnSync } from 'node:child_process'
import { CACHE_ROOT } from './config.js'

/**
 * Result returned after preparing an ephemeral dependency cache workspace.
 */
export interface CacheResult {
  /** Path to the workspace cache directory. */
  workspaceCacheDir: string
  /** Path to the specific dependency set cache directory. */
  cacheDir: string
  /** Path to the node_modules directory within cache. */
  nodeModulesPath: string
  /** Whether a new npm install was triggered. */
  created: boolean
  /** Stash metadata for restoring original node_modules. */
  stash: NodeModulesStash
}

/**
 * Metadata capturing the state of target node_modules before symlinking.
 */
export interface NodeModulesStash {
  /** 'fresh' if node_modules didn't exist, 'link' if it was a symlink, 'dir' if real directory. */
  kind: 'fresh' | 'link' | 'dir'
  /** Original symlink target if kind === 'link'. */
  originalTarget?: string
  /** Backup directory path if kind === 'dir'. */
  backupPath?: string
}

/**
 * Options for cache preparation.
 */
export interface PrepareCacheOptions {
  /** If true, suppresses npm install stdout output. */
  quiet?: boolean
  /** Extra flags passed to npm install. */
  npmArgs?: string[]
  /** If true, prints internal step logs to stderr. */
  debug?: boolean
}

/**
 * Generates a short 16-character SHA-256 hex digest.
 *
 * @param s - Input string to hash.
 * @returns 16-character hex hash string.
 */
function shortHash(s: string): string {
  return createHash('sha256').update(s).digest('hex').slice(0, 16)
}

/**
 * Hashes absolute path of workspace directory to isolate caches per project.
 *
 * @param workspaceDir - Directory path of the workspace.
 * @returns 16-character workspace hash.
 */
export function workspaceHash(workspaceDir: string): string {
  return shortHash(path.resolve(workspaceDir))
}

/**
 * Hashes a list of package specifiers deterministically.
 *
 * @param withList - List of package specifiers.
 * @returns 16-character dependencies hash.
 */
export function depsHash(withList: string[]): string {
  return shortHash([...withList].sort().join('\u0000'))
}

/**
 * Generates a cache key partitioning dependencies and npm flags.
 *
 * @param withList - Package specifiers.
 * @param npmArgs - Extra npm flags.
 * @returns 16-character cache key.
 */
export function cacheKey(withList: string[], npmArgs: string[]): string {
  return shortHash([depsHash(withList), ...npmArgs].join('\u0000'))
}

/**
 * Computes path to run lock file for target directory.
 *
 * @param targetDir - Target directory path.
 * @returns Lock file path inside cache.
 */
function lockPathFor(targetDir: string): string {
  return path.join(CACHE_ROOT, workspaceHash(targetDir), 'run.lock')
}

/**
 * Acquires a non-blocking execution lock for target directory.
 * Recovers stale locks if process PID is dead.
 *
 * @param targetDir - Directory path to lock.
 * @returns A promise resolving to a release function.
 */
export async function acquireRunLock(targetDir: string): Promise<() => Promise<void>> {
  const lockPath = lockPathFor(targetDir)
  await fs.mkdir(path.dirname(lockPath), { recursive: true })

  for (let attempt = 0; attempt < 2; attempt++) {
    let fd: fs.FileHandle | null = null
    try {
      fd = await fs.open(lockPath, 'wx')
      await fd.writeFile(`${process.pid}\n`)
      let closed = false
      return async () => {
        if (closed) return
        closed = true
        await fd?.close().catch(() => {})
        await fs.unlink(lockPath).catch(() => {})
      }
    } catch (err) {
      const code = (err as NodeJS.ErrnoException).code
      if (code !== 'EEXIST') throw err
      const owner = (await fs.readFile(lockPath, 'utf8').catch(() => '')).trim()
      if (isPidAlive(Number(owner))) {
        throw new Error(
          `otro ntsx ya está corriendo en este directorio (pid ${owner}). Espera a que termine o elimina ${lockPath}`
        )
      }
      await fs.unlink(lockPath).catch(() => {})
    }
  }
  throw new Error(`no se pudo adquirir el run lock: ${lockPath}`)
}

/**
 * Checks if process with given PID is alive.
 *
 * @param pid - Process ID.
 * @returns True if process is alive, false otherwise.
 */
function isPidAlive(pid: number): boolean {
  if (!Number.isInteger(pid) || pid <= 0) return false
  try {
    process.kill(pid, 0)
    return true
  } catch (err) {
    return (err as NodeJS.ErrnoException).code === 'EPERM'
  }
}

/**
 * Computes cache directory path for a workspace and dependency set.
 *
 * @param workspaceDir - Project directory.
 * @param withList - Package specifiers.
 * @param npmArgs - Optional npm flags.
 * @returns Absolute cache path.
 */
export function cacheDirFor(workspaceDir: string, withList: string[], npmArgs: string[] = []): string {
  return path.join(CACHE_ROOT, workspaceHash(workspaceDir), cacheKey(withList, npmArgs))
}

/**
 * Prepares dependency cache workspace and symlinks node_modules into target.
 *
 * @param withList - Package specifiers.
 * @param targetDir - Workspace directory.
 * @param opts - Preparation options.
 * @returns Cache setup result and stash state.
 */
export async function prepareCache(
  withList: string[],
  targetDir: string,
  opts: PrepareCacheOptions = {}
): Promise<CacheResult> {
  const wsHash = workspaceHash(targetDir)
  const workspaceCacheDir = path.join(CACHE_ROOT, wsHash)
  const cacheDir = path.join(workspaceCacheDir, cacheKey(withList, opts.npmArgs ?? []))
  const cacheNodeModules = path.join(cacheDir, 'node_modules')

  debugLog(opts.debug, `workspace cache: ${workspaceCacheDir}`)
  debugLog(opts.debug, `deps cache: ${cacheDir}`)
  await fs.mkdir(workspaceCacheDir, { recursive: true })
  await fs.mkdir(cacheDir, { recursive: true })

  let installed = false
  try {
    await fs.access(path.join(cacheDir, 'package.json'))
    await fs.access(cacheNodeModules)
    installed = true
  } catch {
    installed = false
  }

  let created = false
  if (!installed) {
    debugLog(opts.debug, `npm install: ${withList.join(', ')} (${cacheDir})`)
    await writePackageJson(cacheDir, withList)
    runNpmInstall(cacheDir, opts.quiet ?? false, opts.npmArgs ?? [], opts.debug ?? false)
    created = true
  } else {
    debugLog(opts.debug, 'deps already installed in cache (skipping npm install)')
  }

  const nodeModulesPath = path.join(targetDir, 'node_modules')
  const stash = await stashNodeModules(nodeModulesPath)
  debugLog(
    opts.debug,
    `node_modules en ${targetDir}: ${
      stash.kind === 'fresh'
        ? 'no existía (fresh)'
        : stash.kind === 'link'
        ? `symlink → ${stash.originalTarget}`
        : `apartado en ${stash.backupPath}`
    }`
  )
  try {
    await fs.symlink(cacheNodeModules, nodeModulesPath, 'dir')
  } catch (err) {
    await restoreNodeModules(nodeModulesPath, stash)
    throw err
  }
  debugLog(opts.debug, `symlink ${nodeModulesPath} → ${cacheNodeModules}`)

  return { workspaceCacheDir, cacheDir, nodeModulesPath: cacheNodeModules, created, stash }
}

/**
 * Helper to write debug logs to stderr when debug mode is enabled.
 *
 * @param enabled - Debug flag.
 * @param msg - Debug log message.
 */
function debugLog(enabled: boolean | undefined, msg: string): void {
  if (enabled) process.stderr.write(`ntsx: [debug] ${msg}\n`)
}

/**
 * Stashes target node_modules before symlinking ephemeral dependencies.
 *
 * @param p - Absolute path to node_modules in target directory.
 * @returns Stash state description.
 */
async function stashNodeModules(p: string): Promise<NodeModulesStash> {
  try {
    const st = await fs.lstat(p)
    if (st.isSymbolicLink()) {
      const originalTarget = await fs.readlink(p)
      if (originalTarget.startsWith(CACHE_ROOT)) {
        await fs.unlink(p).catch(() => {})
        return { kind: 'fresh' }
      }
      await fs.unlink(p)
      return { kind: 'link', originalTarget }
    }
    if (st.isDirectory()) {
      const backupPath = path.join(path.dirname(p), `.ntsx-${process.pid}-${Date.now()}.bak`)
      await fs.rename(p, backupPath)
      return { kind: 'dir', backupPath }
    }
  } catch {
    // File does not exist
  }
  return { kind: 'fresh' }
}

/**
 * Restores original target node_modules from stash state.
 *
 * @param p - Path to target node_modules.
 * @param stash - Stash metadata.
 */
export async function restoreNodeModules(p: string, stash: NodeModulesStash): Promise<void> {
  if (stash.kind === 'link' && stash.originalTarget) {
    await fs.unlink(p).catch(() => {})
    try {
      await fs.symlink(stash.originalTarget, p, 'dir')
    } catch (err) {
      warn(`no se pudo restaurar el symlink original ${stash.originalTarget} en ${p}: ${errMsg(err)}`)
    }
  } else if (stash.kind === 'dir' && stash.backupPath) {
    await fs.unlink(p).catch(() => {})
    try {
      await fs.rename(stash.backupPath, p)
    } catch (err) {
      warn(`no se pudo restaurar tu node_modules real desde ${stash.backupPath}: ${errMsg(err)}`)
    }
  } else {
    await fs.unlink(p).catch(() => {})
  }
}

/**
 * Prints warning message to stderr.
 *
 * @param msg - Warning message.
 */
function warn(msg: string): void {
  process.stderr.write(`ntsx: warning: ${msg}\n`)
}

/**
 * Formats unknown error value to string error message.
 *
 * @param err - Unknown thrown error.
 * @returns Error message string.
 */
function errMsg(err: unknown): string {
  return err instanceof Error ? err.message : String(err)
}

/**
 * Parses package specifier into [name, version].
 *
 * @param spec - Package specifier string.
 * @returns Tuple of [package name, version requirement].
 */
function parseSpec(spec: string): [string, string] {
  if (spec.startsWith('@')) {
    const at = spec.indexOf('@', 1)
    if (at === -1) return [spec, '*']
    return [spec.slice(0, at), spec.slice(at + 1)]
  }
  const at = spec.indexOf('@')
  if (at === -1) return [spec, '*']
  return [spec.slice(0, at), spec.slice(at + 1)]
}

/**
 * Writes package.json in cache directory with specified dependencies.
 *
 * @param dir - Cache directory path.
 * @param withList - List of package specifiers.
 */
async function writePackageJson(dir: string, withList: string[]): Promise<void> {
  const dependencies: Record<string, string> = {}
  for (const spec of withList) {
    const [name, version] = parseSpec(spec)
    dependencies[name] = version
  }
  const packageJson = {
    name: 'ntsx-cache-' + depsHash(withList),
    type: 'module',
    private: true,
    dependencies,
  }
  await fs.writeFile(path.join(dir, 'package.json'), JSON.stringify(packageJson, null, 2) + '\n')
}

/**
 * Executes npm install in isolated cache directory.
 *
 * @param dir - Cache directory path.
 * @param quiet - If true, suppresses stdout.
 * @param npmArgs - Extra npm flags.
 * @param debug - Debug mode flag.
 */
function runNpmInstall(dir: string, quiet: boolean, npmArgs: string[], debug = false): void {
  const flags = [...npmArgs]
  const cmd = ['npm', 'install', '--no-audit', '--no-fund', ...flags]
  debugLog(debug, `exec: ${cmd.join(' ')}`)
  const res = spawnSync(cmd[0], cmd.slice(1), {
    cwd: dir,
    stdio: ['pipe', debug || !quiet ? 'inherit' : 'ignore', 'pipe'],
    env: cleanNpmEnv(process.env),
    encoding: 'utf8',
  })
  if (res.status !== 0) {
    const detail = (res.stderr || '').trim().split('\n').slice(0, 8).join('\n')
    throw new Error(`npm install failed (exit ${res.status ?? 'unknown'}) in ${dir}\n${detail}`)
  }
  debugLog(debug, 'npm install ok')
}

/**
 * Sanitizes environment variables to prevent inheriting npm_config_* variables.
 *
 * @param base - Process environment variables object.
 * @returns Cleaned process environment object.
 */
function cleanNpmEnv(base: NodeJS.ProcessEnv): NodeJS.ProcessEnv {
  const env: NodeJS.ProcessEnv = { ...base }
  for (const key of Object.keys(env)) {
    if (/^npm_config_/i.test(key)) delete env[key]
  }
  return env
}

File: rust/src/cache.rs
use crate::config::get_cache_root;
use std::collections::BTreeMap;
use std::fs;
use std::io::{self, Write};
use std::path::{Path, PathBuf};
use std::process::Command;

pub struct CacheStats {
    pub exists: bool,
    pub workspace_count: usize,
    pub size_bytes: u64,
}

pub struct CleanCacheResult {
    pub cleared: bool,
    pub size_freed: u64,
}

pub enum StashKind {
    Fresh,
    Link(String),
    Dir(PathBuf),
}

pub struct NodeModulesStash {
    pub kind: StashKind,
}

pub fn fmt_bytes(bytes: u64) -> String {
    if bytes < 1024 {
        format!("{} B", bytes)
    } else if bytes < 1024 * 1024 {
        format!("{:.1} KB", bytes as f64 / 1024.0)
    } else if bytes < 1024 * 1024 * 1024 {
        format!("{:.1} MB", bytes as f64 / (1024.0 * 1024.0))
    } else {
        format!("{:.2} GB", bytes as f64 / (1024.0 * 1024.0 * 1024.0))
    }
}

pub fn short_hash(input: &str) -> String {
    let mut hash: u64 = 0xcbf2_9ce4_8422_2325;
    for byte in input.bytes() {
        hash ^= byte as u64;
        hash = hash.wrapping_mul(0x0100_0000_01b3);
    }
    format!("{:016x}", hash)
}

pub fn workspace_hash(target_dir: &Path) -> String {
    short_hash(&target_dir.to_string_lossy())
}

pub fn deps_hash(with_list: &[String]) -> String {
    let mut sorted = with_list.to_vec();
    sorted.sort();
    short_hash(&sorted.join("\0"))
}

pub fn cache_key(with_list: &[String], npm_args: &[String]) -> String {
    let mut combined = vec![deps_hash(with_list)];
    combined.extend(npm_args.iter().cloned());
    short_hash(&combined.join("\0"))
}

fn dir_size(path: &Path) -> u64 {
    let mut total = 0;
    if let Ok(entries) = fs::read_dir(path) {
        for entry in entries.flatten() {
            let p = entry.path();
            if let Ok(meta) = fs::symlink_metadata(&p) {
                if meta.is_dir() {
                    total += dir_size(&p);
                } else {
                    total += meta.len();
                }
            }
        }
    }
    total
}

pub async fn cache_stats() -> Result<CacheStats, Box<dyn std::error::Error>> {
    let root = get_cache_root();
    if !root.exists() {
        return Ok(CacheStats {
            exists: false,
            workspace_count: 0,
            size_bytes: 0,
        });
    }

    let mut workspace_count = 0;
    if let Ok(entries) = fs::read_dir(&root) {
        for entry in entries.flatten() {
            if entry.path().is_dir() {
                workspace_count += 1;
            }
        }
    }

    let size_bytes = dir_size(&root);
    Ok(CacheStats {
        exists: true,
        workspace_count,
        size_bytes,
    })
}

pub async fn clean_cache(force: bool) -> Result<CleanCacheResult, Box<dyn std::error::Error>> {
    let stats = cache_stats().await?;
    if !stats.exists {
        return Ok(CleanCacheResult {
            cleared: false,
            size_freed: 0,
        });
    }

    let size_freed = stats.size_bytes;
    let cache_root = get_cache_root();

    let confirmed = if force {
        true
    } else {
        print!(
            "Delete cache {} ({})? [y/N] ",
            cache_root.display(),
            fmt_bytes(size_freed)
        );
        io::stdout().flush()?;
        let mut input = String::new();
        io::stdin().read_line(&mut input)?;
        let trimmed = input.trim();
        trimmed.eq_ignore_ascii_case("y") || trimmed.eq_ignore_ascii_case("yes")
    };

    if !confirmed {
        return Ok(CleanCacheResult {
            cleared: false,
            size_freed,
        });
    }

    if cache_root.exists() {
        fs::remove_dir_all(&cache_root)?;
    }

    Ok(CleanCacheResult {
        cleared: true,
        size_freed,
    })
}

pub async fn stash_node_modules(p: &Path) -> Result<NodeModulesStash, Box<dyn std::error::Error>> {
    if let Ok(meta) = fs::symlink_metadata(p) {
        if meta.file_type().is_symlink() {
            if let Ok(target) = fs::read_link(p) {
                let cache_root_str = get_cache_root().to_string_lossy().to_string();
                if target.to_string_lossy().contains(&cache_root_str) {
                    let _ = fs::remove_file(p);
                    return Ok(NodeModulesStash {
                        kind: StashKind::Fresh,
                    });
                }
                let _ = fs::remove_file(p);
                return Ok(NodeModulesStash {
                    kind: StashKind::Link(target.to_string_lossy().to_string()),
                });
            }
        } else if meta.is_dir() {
            let backup_path = p
                .parent()
                .unwrap_or_else(|| Path::new("."))
                .join(format!(".ntsx-{}.bak", std::process::id()));
            let _ = fs::rename(p, &backup_path);
            return Ok(NodeModulesStash {
                kind: StashKind::Dir(backup_path),
            });
        }
    }
    Ok(NodeModulesStash {
        kind: StashKind::Fresh,
    })
}

pub async fn restore_node_modules(
    p: &Path,
    stash: NodeModulesStash,
) -> Result<(), Box<dyn std::error::Error>> {
    let _ = fs::remove_file(p);
    let _ = fs::remove_dir_all(p);

    match stash.kind {
        StashKind::Link(original_target) => {
            #[cfg(unix)]
            {
                use std::os::unix::fs::symlink;
                let _ = symlink(original_target, p);
            }
            #[cfg(windows)]
            {
                use std::os::windows::fs::symlink_dir;
                let _ = symlink_dir(original_target, p);
            }
        }
        StashKind::Dir(backup_path) => {
            if backup_path.exists() {
                let _ = fs::rename(backup_path, p);
            }
        }
        StashKind::Fresh => {}
    }
    Ok(())
}

pub fn parse_spec(spec: &str) -> (String, String) {
    if let Some(stripped) = spec.strip_prefix('@') {
        if let Some(at) = stripped.find('@') {
            let name = format!("@{}", &stripped[..at]);
            let ver = &stripped[at + 1..];
            return (name, ver.to_string());
        } else {
            return (spec.to_string(), "*".to_string());
        }
    }
    if let Some(at) = spec.find('@') {
        let (name, ver) = spec.split_at(at);
        (name.to_string(), ver[1..].to_string())
    } else {
        (spec.to_string(), "*".to_string())
    }
}

pub async fn prepare_cache(
    with_list: &[String],
    target_dir: &Path,
    quiet: bool,
    npm_args: &[String],
    _debug: bool,
) -> Result<NodeModulesStash, Box<dyn std::error::Error>> {
    let ws_hash = workspace_hash(target_dir);
    let ws_cache_dir = get_cache_root().join(ws_hash);
    let dep_cache_dir = ws_cache_dir.join(cache_key(with_list, npm_args));
    let cache_node_modules = dep_cache_dir.join("node_modules");

    fs::create_dir_all(&ws_cache_dir)?;
    fs::create_dir_all(&dep_cache_dir)?;

    let pkg_json_path = dep_cache_dir.join("package.json");
    if !pkg_json_path.exists() || !cache_node_modules.exists() {
        let mut deps = BTreeMap::new();
        for spec in with_list {
            let (name, version) = parse_spec(spec);
            deps.insert(name, version);
        }

        let pkg_manifest = serde_json::json!({
            "name": format!("ntsx-cache-{}", deps_hash(with_list)),
            "type": "module",
            "private": true,
            "dependencies": deps
        });

        fs::write(&pkg_json_path, serde_json::to_string_pretty(&pkg_manifest)?)?;

        let mut npm_cmd = Command::new("npm");
        npm_cmd
            .arg("install")
            .arg("--no-audit")
            .arg("--no-fund")
            .current_dir(&dep_cache_dir);
        for extra in npm_args {
            npm_cmd.arg(extra);
        }

        if quiet {
            npm_cmd
                .stdout(std::process::Stdio::null())
                .stderr(std::process::Stdio::null());
        }

        let output = npm_cmd.output()?;
        if !output.status.success() {
            let stderr_str = String::from_utf8_lossy(&output.stderr);
            let detail = stderr_str
                .trim()
                .lines()
                .take(8)
                .collect::<Vec<_>>()
                .join("\n");
            return Err(format!("Invalid package spec (expected pkg, pkg@version, @scope/pkg, or @scope/pkg@version)\n\n{detail}").into());
        }
    }

    let target_node_modules = target_dir.join("node_modules");
    let stash = stash_node_modules(&target_node_modules).await?;

    #[cfg(unix)]
    {
        use std::os::unix::fs::symlink;
        symlink(&cache_node_modules, &target_node_modules)?;
    }
    #[cfg(windows)]
    {
        use std::os::windows::fs::symlink_dir;
        symlink_dir(&cache_node_modules, &target_node_modules)?;
    }

    Ok(stash)
}
